aten/src/ATen/core/autocast/register_autocast_ops.cpp

```diff
@@ -0,0 +1,528 @@
+  }
+}
+
+// Overload to process TensorLists
+std::vector<Tensor> cached_cast(at::ScalarType to_type, const TensorList& arg) {
+  std::vector<Tensor> vec;
+  vec.reserve(arg.size());
+  for (const auto& t : arg) {
+    vec.push_back(cached_cast(to_type, t));
+  }
+  return vec;
+}
+
+// Template to catch non-Tensor args.
+template<typename T> inline T
+cached_cast(at::ScalarType to_type, T arg) {
+  return arg;
+}
+
+/*******************************************************************************
+Logic to apply cached casting to signatures with Tensors and a single TensorList
+The functions here exist only to support CastPolicy::*_with_tensorlist,
+which store a local vector to preserve the lifetime of data that's viewed by
+TensorLists further down the call chain.
+*******************************************************************************/
+
+// helper to pick out the TensorList arg and create a casted vector
+template<typename Arg0, typename... Args> inline std::vector<Tensor>
+get_casted_vector(at::ScalarType to_type, Arg0 arg0, Args... args) {
+  return get_casted_vector(args...);
+}
+
+// Specialization for the TensorList itself
+template<typename... Args> inline std::vector<Tensor>
+get_casted_vector(at::ScalarType to_type, const TensorList& list, Args... args) {
+  return cached_cast(to_type, list);
+}
+
+// Overload to catch Tensor args
+inline Tensor
+cached_cast_with_vector(at::ScalarType to_type, const std::vector<Tensor>& casted_vector, const Tensor& arg) {
+  return cached_cast(to_type, arg);
+}
+
+// Overload to catch the TensorList arg and swap in our vector
+inline const std::vector<Tensor>&
+cached_cast_with_vector(at::ScalarType to_type, const std::vector<Tensor>& casted_vector, const TensorList& arg) {
+  return casted_vector;
+}
+
+// Template to catch other args.
+template<typename T> inline T
+cached_cast_with_vector(at::ScalarType to_type, const std::vector<Tensor>& casted_vector, T arg) {
+  return arg;
+}
+
+/*******************************************************
+Logic to flip an output dtype flag.
+Keep it simple for now by assuming only one such flag is
+present in the argument list.  If I ever need a function
+with more than flag I'll figure out something else.
+The policy is:
+If the user has explicity specified a dtype, respect it.
+Otherwise, set it to the autocast type.
+********************************************************/
+
+// Overload to catch dtype flags
+inline c10::optional<ScalarType> set_opt_dtype(at::ScalarType to_type, const c10::optional<ScalarType>& dtype) {
+  return dtype.has_value() ? dtype : to_type;
+}
+
+// Template to catch other args
+template<typename T>
+inline T set_opt_dtype(at::ScalarType to_type, T arg) {
+  return arg;
+}
+
+/********************************************************************************************************
+Templates to provide wrapper functions
+
+I'm copying the pattern used in core/boxing/kernel_function.h to extract args and return type.
```

Comment:
@smessmer, you're probably going to wanna take a look at this.

Reply:
It's kind of interesting that AMP is implemented by metaprogramming out the wrappers, as opposed to using the boxed interface. Not making a judgment about this code here, but in the original design, boxing should have been a way to write this out in a non-templated way. It's interesting to see that it's not the case here.

Reply:
Actually I didn't consider that.  It makes sense that my argument forwarding ended up mimicking the boxed interface.  I think I could rewrite the code to use a boxed interface, but many (like half) of the kernels I'm registering for are registered as "kernelUnboxedOnly" (blindly imitating how they're registered in VariableType*.cpp).  Maybe that means I can't use the boxed interface, i'd have to look more closely.

Reply:
The boxed interface would be (1) slower but (2) result in less generated code. Maybe it would be easier to read although that's an eye of the beholder thing. I think as it stands is fine.